cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/LineageInfo.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.spi.metadata;
+
+import io.cdap.cdap.api.annotation.Beta;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Lineage information for a program run.
+ */
+@Beta
+public class LineageInfo {
+
+  private final Set<Dataset> sources;
+  private final Set<Dataset> targets;
+  private final Map<Dataset, Set<Dataset>> sourceToTargets;
+  private final Map<Dataset, Set<Dataset>> targetToSources;
+  private final Long startTimeMs;
+  private final Long endTimeMs;
+
+  private LineageInfo(Set<Dataset> sources, Set<Dataset> targets, Map<Dataset, Set<Dataset>> sourceToTargets,
+                      Map<Dataset, Set<Dataset>> targetToSources, @Nullable Long startTimeMs,
```

Comment:
When can start and end time be null? For end time, I understand it can be null for streaming pipeline, when can start time be null?

Reply:
Yes for streaming pipelines `endTime` can be null. `startTime` is not likely to be null anytime, but since SPI implementations need not use this value to be set, keeping it Nullable